src/lib/efthrm/tcp_helper_resource.c

```diff
@@ -8413,24 +8417,36 @@ static void oo_inject_packets_work(struct work_struct* work)
   ci_netif* ni = &data->trs->netif;
   ci_ip_pkt_fmt* pkt;
   int netif_is_locked;
+  unsigned n;
 
   /* Part one: inject all packets to the kernel */
-  for( pkt = PKT_CHK(ni, data->pkt_head); ; pkt = PKT_CHK(ni, pkt->next) ) {
+  for( pkt = PKT_CHK(ni, data->pkt_head), n = 0;
+       n < data->n_pks;
+       pkt = PKT_CHK(ni, pkt->next), n++ ) {
     /* No need to check the return value here.  If the function fails, the
      * packet is dropped, and a counter is incremented. */
     oo_inject_packet_kernel(ni, pkt);
 
     if( OO_PP_IS_NULL(pkt->next) )
       break;
+    ci_assert_lt(n, data->n_pks);
+    if( n == data->n_pks )
```

Comment:
I guess this would be a tighter check:
```suggestion
    if( n >= data->n_pks )
```
Could you clarify: when does it occur? `n` and `n_pks` both seem kernelspace-only variables and the for-loop statement prevents us from getting here.

Presumably, a list of `pkt` can be corrupted, but the `n < data->n_pks;` silently saves us?

Reply:
Yes, the list of packets can be corrupted (because the packets are UL-mapped), and we should limit the loop with something non-shared-with-UL.  That's one of the ideas of this patch.  Such an idea already happens in some places in Onload.

Relevant part of the commit log:
```
Also add a guard to avoid dead loop in kernel when shared state
is corrupted.
```

Reply:
Thanks @ol-alexandra. Does the assertion and `if` ever occur though? Both `n` and `data->n_pks` are non-shared.

Reply:
In theory: the last packet is corrupted , and `pkt->next` is non-null.  So we don't break at
```
   if( OO_PP_IS_NULL(pkt->next) )
      break;
```
because of corruption.  Then we hit assertion (assertions traditionally assume well-behaving UL) and the condition.  The condition definitely needs another break; I'll fix.

Reply:
Thanks, that's how I understand it too, but what puzzles me is:
```c
for( ... , n = 0;
     n < data->n_pkts; /* We shouldn't enter the loop if n >= data->n_pkts. */
     ... , n++ ) {

  /* Nothing changes @n and @data (including @data->n_pkts). */

  /* I can't understand why this can occur even if UL fiddles with shared memory. */
  ci_assert_lt(n, data->n_pkts);
  if( n >= data->n_pkts ) {
    ...
    break;
  }
}
```